int-set/src/input_bit_stream.rs

```diff
@@ -1,137 +1,161 @@
+    type Item = u32;
+    fn next(&mut self) -> Option<Self::Item> {
+        match BF {
+            2 | 4 => {
+                let mask = (1 << BF) - 1;
+                let byte = self.data.get(self.byte_index)?;
+                let val = (*byte as u32 & (mask << self.sub_index)) >> self.sub_index;
+                self.sub_index = (self.sub_index + BF as u32) % 8;
+                if self.sub_index == 0 {
+                    self.byte_index += 1;
+                }
+                Some(val)
+            }
+            8 => {
+                let r = self.data.get(self.byte_index).map(|v| *v as u32)?;
+                self.byte_index += 1;
+                Some(r)
+            }
 
-    /// Reads the next two bits and returns the corresponding branch factor.
-    ///
-    /// See: <https://w3c.github.io/IFT/Overview.html#sparse-bit-set-decoding>
-    /// Returns None if the stream does not have enough remaining bits.
-    pub fn read_branch_factor(&mut self) -> Option<u8> {
-        let bit_0 = self.read_bit()?;
-        let bit_1 = self.read_bit()?;
-
-        match (bit_1, bit_0) {
-            (false, false) => Some(2),
-            (false, true) => Some(4),
-            (true, false) => Some(8),
-            (true, true) => Some(32),
+            32 => {
+                let b1 = self.data.get(self.byte_index).map(|v| *v as u32)?;
+                let b2 = self.data.get(self.byte_index + 1).map(|v| *v as u32)?;
+                let b3 = self.data.get(self.byte_index + 2).map(|v| *v as u32)?;
+                let b4 = self.data.get(self.byte_index + 3).map(|v| *v as u32)?;
+                self.byte_index += 4;
+                Some(b1 | (b2 << 8) | (b3 << 16) | (b4 << 24))
+            }
+            _ => panic!("Unsupported branch factor."),
         }
     }
+}
 
-    /// Reads the next 5 bits and returns the corresponding height value.
+impl<'a, const BF: u8> InputBitStream<'a, BF> {
+    /// Decodes and returns the branch factor and height encoded in the header byte.
     ///
     /// See: <https://w3c.github.io/IFT/Overview.html#sparse-bit-set-decoding>
     /// Returns None if the stream does not have enough remaining bits.
-    pub fn read_height(&mut self) -> Option<u8> {
-        let mut val = 0u8;
-        for index in 0..5 {
-            if self.read_bit()? {
-                val |= 1u8 << index;
-            }
-        }
-
-        Some(val)
-    }
-
-    /// Returns the value of the next bit, or None if there are no more bits left.
-    ///
-    /// Bits are read from least signifcant to most in the current byte.
-    pub fn read_bit(&mut self) -> Option<bool> {
-        let byte = self.data.get(self.byte_index)?;
-        let mask = 1u8 << self.bit_index;
-        let bit_value = byte & mask != 0;
-
-        self.move_to_next();
-
-        Some(bit_value)
-    }
-
-    /// Skips the next bit in the stream.
-    pub fn skip_bit(&mut self) {
-        self.move_to_next();
+    #[allow(clippy::unusual_byte_groupings)]
```

Comment:
when opting out of a clippy lint I find it helpful to include a little comment explaining why.